### PR TITLE
PFW-1334 Move E-motor on FSENSOR error retry

### PR DESCRIPTION
**Pull request by gudnimg**
JIRA ticket: https://dev.prusa3d.com/browse/PFW-1334

When a FSENSOR error happens, the first step for the MK3S should be to stop any loading by stopping the E-motor.

From this point, the buttons should determine what the MK3S does next.

### CODE REVIEW

**Comment by gudnimg** [comment]
~~The first commit fixes the scenario for `FSENSOR DIDNT TRIGGER`, because the Retry button correctly restarts the state machine by unloading the filament and repeating the load. Specifically this is for loading, not unloading.~~

EDIT: Correction, the above commit only stops the motor from moving endlessly after an error appears. Unloading is not triggered and will be added in a later commit.

I have yet to verify the other errors.

**Comment by gudnimg** [comment]
* `FSENSOR TOO EARLY` is triggered during loading. If the FINDA is set to 1 during load as normal but the filament sensor is set 1 too quickly then this error is raised. A retry should unload to FINDA.

* `FSENSOR DIDNT GO OFF` is a bit more tricky. Because here we are not loading, but unloading. I am able to trigger this error if the retry for `FSENSOR TOO EARLY` fails. For whatever reason when retry is selected here, there is no U0 sent to the MMU like with `FSENSOR DIDNT TRIGGER`. 🤔

**Comment by gudnimg** [comment]
I will likely now need to distinguish between the normal call to `unload()` and a unload retry since they do not behave the same.

**Comment by gudnimg** [comment]
I'm marking this as ready for review, it should be safe to use now.

**Comment by gudnimg** [comment]
Rebased branch onto mk3-mmu

File: Firmware/mmu2.cpp [hunk 1]
```diff
@@ -765,6 +782,26 @@ void MMU2::OnMMUProgressMsg(ProgressCode pc){
...
+                break;
+            }
+
+            // This is intended to handle Retry option on MMU error screen
+            // MMU sends P3 progress code during Query, and if filament is stuck
+            // in the gears, the MK3S needs to move e-axis as well.
+            st_synchronize();
+            unloadFilamentStarted = true;
+            // Unload slowly while MMU is initalising its axis
+            current_position[E_AXIS] -= 10.0f;
```
Comment (3d-gussner) [inline Firmware/mmu2.cpp]: Why do we have hard coded values?

File: Firmware/mmu2.cpp [hunk 2]
```diff
@@ -765,6 +782,26 @@ void MMU2::OnMMUProgressMsg(ProgressCode pc){
...
+            plan_buffer_line_curposXYZE(10.0f);
+            st_synchronize();
+            // Now do a fast unload in sync with the MMU
+            current_position[E_AXIS] -= 427.0f - 42.85f - 20.0f; // Roughly same distance as MMU plans
```
Comment (3d-gussner) [inline Firmware/mmu2.cpp]: Same as above

File: Firmware/mmu2.cpp [hunk 3]
```diff
@@ -777,6 +814,19 @@ void MMU2::OnMMUProgressMsg(ProgressCode pc){
     } else {
         // Act accordingly - every status change (even the same state)
         switch (pc) {
+        case ProgressCode::UnloadingToFinda:
+            if (unloadFilamentStarted && !blocks_queued()) { // Only plan a move if there is no move ongoing
+                
+                if (mmu2.FindaDetectsFilament() == 1)
+                { // Keep moving the E-motor until and error happens or FINDA untriggers
+                    current_position[E_AXIS] -= 6.0f;
```
Comment (3d-gussner) [inline Firmware/mmu2.cpp]: Same as above

**Comment by vintagepc** [comment]
I agree with @3d-gussner - if we must have magic numbers, they should at least be pulled to the top of the file with a static constexpr rather than scattered throughout the code. 

Of course, an improved solution with shared headers etc would be preferred.

**Comment by gudnimg** [comment]
I will hopefully add a new commit later today with global constants and some documentation about them. While I was looking into documenting the constants I found that this is probably incorrect: `-= 427.0f - 42.85f - 20.0f` it should be `-= 427.0f + 42.85f + 20.0f`. I can verify this later today also.

This is basically this line from the MMU FW:  `mpu::pulley.PlanMove(-config::defaultBowdenLength - config::feedToFinda - config::filamentMinLoadedToMMU, config::pulleyUnloadFeedrate);`

**Comment by gudnimg** [comment]
@3d-gussner @vintagepc I've pulled everything into constants. The variable names are a bit long I know 🙈 I tried to keep them descriptive and general.

The constant `-= 427.0f - 42.85f - 20.0f` is now corrected to `-= 427.0f + 42.85f + 20.0f` and it looks much better when executed. The MK3S is more in sync with the MMU. A new problem arose, `427.0f + 42.85f + 20.0f` is now considered too long of an extrusion so I've split this constant into two to keep the firmware happy.

**Comment by gudnimg** [comment]
I've done some more testing and cannot see any issues with tool changes (during printing). I can also confirm the changes in the code are not blocking the MMU communication. So if the testers see the communication drop on this build, it must be something else unrelated to these changes.

The only issue left looking into is the reported layer shifts. I can confirm there aren't any gcodes being executing while loading or unloading (even during a retry). But one idea I have is we are skipping gcodes perhaps. I'm currently looking into how to log that to confirm it.

**Review by 3d-gussner — APPROVED** [review]

**Pull request merged**